sqlglot/optimizer/annotate_types.py

```diff
@@ -368,7 +368,7 @@ def annotate_scope(self, scope: Scope) -> None:
         # Iterate through all the expressions of the current scope in post-order, and annotate
         self._annotate_expression(scope.expression, scope, selects)
 
-        if self.schema.dialect == "bigquery" and isinstance(scope.expression, exp.Query):
+        if Dialect.get_or_raise(self.schema.dialect).QUERY_RESULTS_ARE_STRUCTS and isinstance(scope.expression, exp.Query):
```

Comment:
it'd be better if schema.dialect was the dialect object that way you don't need to call get_or_raise, can we adjust this?